### PR TITLE
3731 bulk share bottleneck

### LINKED ISSUE #3731: Access levels: bulk selection and share takes a long time to load (13 pts)
Selecting a relatively small amount of entities, ie. 150 and configuring the sharing options takes several seconds to display the dialog. There is a performance bottleneck somewhere.

### PR DESCRIPTION
**Pull request by LaszloKecskes**
fixes #3731 

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Review by mfacar — CHANGES_REQUESTED** [review]
The code LGTM!, I left a couple of comments about minor things. I'll put the results of QA just after finishing the tests on the functionality

**Comment by mfacar** [comment]
@LaszloKecskes, QA went pretty well, there's an important reduction in the response time, I think that this improvement can also be applied to the set method in `entitiesPermissions`. Although it's not part of this issue, please can you check it?

File: app/api/entities/entities.js [hunk 1]
```diff
@@ -332,19 +332,30 @@ const validateWritePermissions = (ids, entitiesToUpdate) => {
...
+    if (idFileMap.has(file.entity)) {
+      idFileMap.get(file.entity).push(file);
+    } else {
+      idFileMap.set(file.entity, [file]);
+    }
+  });
+  const result = entities.map(entity => {
+    const entityFiles = idFileMap.has(entity.sharedId)
+      ? idFileMap.get(entity.sharedId).map(file => ({ ...file }))
+      : [];
```
[the remark this answers is not in the document]
Comment (LaszloKecskes) [reply app/api/entities/entities.js:352]: That was what I've done first. It breaks a unittest in the search api, due to wrong indexing. What happens is:
1, Have entities that share documents.
2, During the indexing process [this function](https://github.com/huridocs/uwazi/blob/e95450312781d3ece88ac5a57fd4ae5a0aafa383/app/api/search/entitiesIndex.js#L45) is reached, which deletes the fullText from the js object representing the file.
3, If you pass the array (and the objects) directly, the js object will be shared between the entities.
4, The function deletes the fullText on the first entity that has the file.
5, On the second entity, it sees that there is no fullText, and bad information will be passed to elastic, breaking the indexing.

So, in this case, it is intentional that a copy is passed, not the original object. 

I have also thought about leaving a comment here, marking that this is intentional. Now, that you've find this, I think that's a good proof that that will be a necessary comment. :)

File: app/api/entities/specs/entities.spec.js [hunk 2]
```diff
@@ -668,6 +670,28 @@ describe('entities', () => {
...
+      expect(entity.title).toBe(title);
+
+      if (documentFilenames !== null) {
+        expect(entity.documents.length).toBe(documentFilenames.length);
+        entity.documents.forEach((element, index) => {
+          expect(element.filename).toBe(documentFilenames[index]);
+        });
+      } else {
+        expect(entity).not.toHaveProperty('documents');
+      }
```
[the remark this answers is not in the document]
Comment (LaszloKecskes) [reply app/api/entities/specs/entities.spec.js:683]: Sure, will do so.

File: app/api/entities/entities.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfacar) [reply app/api/entities/entities.js:355]: Ohh, good to know, I think if a test is broken changing it, the comment it's not needed. I leave it to your decision :)

**Comment by mfacar** [comment]
> @LaszloKecskes, QA went pretty well, there's an important reduction in the response time, I think that this improvement can also be applied to the set method in `entitiesPermissions`. Although it's not part of this issue, please can you check it?

The team decided to open a new issue (#3804) to tackle this.

**Review by mfacar — APPROVED** [review]

**Pull request merged**